python/ray/data/_internal/pandas_block.py

```diff
@@ -87,7 +92,31 @@ def _table_from_pydict(self, columns: Dict[str, List[Any]]) -> "pandas.DataFrame
 
     def _concat_tables(self, tables: List["pandas.DataFrame"]) -> "pandas.DataFrame":
         pandas = lazy_import_pandas()
-        return pandas.concat(tables, ignore_index=True)
+        from ray.data.extensions.tensor_extension import TensorArray
+
+        df = pandas.concat(tables, ignore_index=True)
+        # Try to convert any ndarray columns to TensorArray columns.
+        # TODO(Clark): Once Pandas supports registering extension types for type
+        # inference on construction, implement as much for NumPy ndarrays and remove
+        # this. See https://github.com/pandas-dev/pandas/issues/41848
+        for col_name, col in df.items():
+            if (
+                col.dtype.type is np.object_
+                and not col.empty
+                and isinstance(col[0], np.ndarray)
+            ):
+                try:
+                    df[col_name] = TensorArray(col)
+                except Exception as e:
+                    global _tensor_cast_failed_warned
+                    if not _tensor_cast_failed_warned:
```

Comment:
Added this guard to keep us from spamming the user with these cast failures, but this is less clear-cut than our epoch warnings and slow read warnings IMO.

Reply:
How about log_every_n() or log_every_n_seconds(), instead of using a global var?

Reply:
Hmm in the common case where this failed cast happens in a UDF map task, this logging will be happening in the workers, not the driver, so we aren't going to be able to have a global throttle on the log spam. This is already going to log in every map task, so all that this global guard does is prevent workers from re-logging for future stages... Even with the global guard, this is what the logs look like on my local machine:
```python
In [4]: ds.map_batches(lambda df: pd.DataFrame({"a": [np.full((2, 2, i), i) for i in df["value"]]}), batch_size=2)
Read->Map_Batches:   0%|                                                                                                                                                                                                                                                                                                                                                                                                                                                           | 0/20 [00:00<?, ?it/s]
(_map_block_nosplit pid=2428375)   self._tensor = np.array([np.asarray(v) for v in values])
(_map_block_nosplit pid=2428375) 2022-07-18 23:48:10,232        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,5) into shape (2,2)                                                                                                                                                                                                                                                                                                           
(_map_block_nosplit pid=2428875) 2022-07-18 23:48:10,226        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,0) into shape (2,2)
(_map_block_nosplit pid=2428381) 2022-07-18 23:48:10,243        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,15) into shape (2,2)
(_map_block_nosplit pid=2428378) 2022-07-18 23:48:10,258        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,25) into shape (2,2)
(_map_block_nosplit pid=2428376) 2022-07-18 23:48:10,248        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,20) into shape (2,2)
(_map_block_nosplit pid=2428377) 2022-07-18 23:48:10,275        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,40) into shape (2,2)
(_map_block_nosplit pid=2428380) 2022-07-18 23:48:10,242        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,10) into shape (2,2)

Out[4]: Dataset(num_blocks=20, num_rows=100, schema={a: object})

(_map_block_nosplit pid=2429207) 2022-07-18 23:48:11,633        WARNING pandas_block.py:114 -- Tried to transparently convert column a to a TensorArray but the conversion failed, leaving column as-is: could not broadcast input array from shape (2,2,85) into shape (2,2)
```

Reply:
not sure if it's feasiable to add a stats counter here to record how many exceptions happen? So it's easier to debug just look at dataset stats, in case we need this information.

Reply:
I think what's useful here is making sure that the user sees the log message at least _once_, rather than giving them a count of how many times it occurs. The hope is that the user will see the log message and will either (1) disregard it because they expect it to fail, e.g. if they are using heterogeneous tensor columns, or (2) change their NumPy arrays so they're compatible with `TensorArray`.

Reply:
I am more thinking that for us (developers) to debug, not for users. Given we have the discussion around the extra overhead it would introduce, thought it would be good to add a counter to catch it, so next time it's easier to debug. But we can decide to do it or not later.

Reply:
@jianoaix I understand that general pattern when the logging is centralized, but how valuable is a per-worker `log_every_n()` or `log_every_n_seconds()` going to be when the logging is going to happen once per stage but for every worker?

Since we don't want to introduce expensive coordination for the sake of log throttling, the options are:
1. Log once per worker per stage (no local logging guard).
2. Log once per worker (current local logging guard).
3. Log every once in a while per worker (your proposed local "log every n (seconds)").

I'm thinking that we want either (1), which gives complete information to the user and allows them to identify stages in which this is happening, or (2), which lets the user know that this problem exists but doesn't spam them with logs, and I'm not sure what value the compromise (3) would give... maybe that would reduce the amount of spam when doing batch inference over a size-constrained actor pool while still periodically letting the user know that this failed cast is happening, but I feel like (2) is still better since (3) could end up spamming the user for a long-running batch inference stage.

Reply:
Then why not log_first_n()?

Reply:
For a long-running service, making the log recurring (not flood for sure) isn't necessarily spam, it may show the problem to users clearer/better than log just once.

Reply:
We have a pattern for this already:

```
if ray.util.log_once("pandas_tensor_warning"):
    logger.warning(...)
```

Let's just use this as the standard way for logging once per process.

Reply:
Ah nice, I wasn't aware that we had a utility for that!

@ericl Do you think that we should log once per worker, or periodically in each worker? I'm leaning towards the latter, so the user is more likely to see these log messages in long-running pipelines even if a lot of worker reuse is happening.

Reply:
If we do we should have a common util. I think either is fine for now.

Reply:
Yep logging periodically is already possible with `ray.util.log_once()`, hardcoded to 60 second debouncing. @jianoaix how does that sound?

Reply:
@jianoaix I started off with using the plain `ray.util.log_once()`, since switching to 60 second debouncing would require either (a) setting a global config that will affect other `ray.util.log_once()` uses, or (b) adding a new common utility.

Reply:
Ok, sounds good. Thanks for looking into this!